Add BACHelp api

stouffers opened the pull request:
Adds new api methods accessible via a new key based auth method. Methods are
targeted for use by the BACHelp ticketing system.
- empSearch: searches for employees by name, returning only fields
  needed by bachelp
- statusChanges: returns a list of employees with certain status changes
  after the given "from" date

Implements simple key based auth by creating a new bachelp authorizing realm
that authenticates using an api key set in app.properties.

Makes significant changes to test config to allow testing of the new auth
realm:
- Properly add shiro filters to WebTest, allowing auth testing using
  MockMvc and incorporating auth for all WebTest subclasses
- Add protected utility methods to WebTest allowing easy authentication
  during tests.
- Add new conf file test.default.properties which is used in TestConfig
  to override properties to values expected in certain tests. Unlike
  test.app.properties, this file is committed.

stouffers committed:
Add api endpoints for bachelp integration

- Add search endpoint returning limited employee info
- Add statusChanges endpoint that gets employees who have had particular
  transactions posted after a given date
- Add postRecordsSince to EmpTransactionDao to retrieve records based on
  post date
- Add posted date to transaction record model

stouffers committed:
Add BACHelp API

Adds new api methods accessible via a new key based auth method. Methods are
targeted for use by the BACHelp ticketing system.
- empSearch: searches for employees by name, returning only fields
  needed by bachelp
- statusChanges: returns a list of employees with certain status changes
  after the given "from" date

Implements simple key based auth by creating a new bachelp authorizing realm
that authenticates using an api key set in app.properties.

Makes significant changes to test config to allow testing of the new auth
realm:
- Properly add shiro filters to WebTest, allowing auth testing using
  MockMvc and incorporating auth for all WebTest subclasses
- Add protected utility methods to WebTest allowing easy authentication
  during tests.
- Add new conf file test.default.properties which is used in TestConfig
  to override properties to values expected in certain tests. Unlike
  test.app.properties, this file is committed.

KevinCaseiras commented:
Hey @stouffers, I can't find the `test.default.properties` file. Was it supposed to be committed or do I need to create it?

stouffers committed:
Fix issue where transaction post date was null

BACHelp status changes api was returning results with null post dates.
This was due to an incomplete copy constructor on the transactioninfo
class. Fixed that and added a test to verify.

stouffers committed:
Add employee lookup endpoint to BACHelp API

- Add GET /api/v1/bachelp/employee/{empId} endpoint for individual employee lookup
- Add proper permission check and exception handling for employee not found
- Change employee search endpoint path from /empSearch to /employee/search
- Add integration tests for both success and error cases

stouffers committed:
Add Resp Ctr Head to BACHelpEmployeeView

Needed for office info in bachelp

stouffers committed:
Add location data to BACHelp Api responses

stouffers committed:
Change params of trans. rec. by post date method

Instead of just a "from" date, make the params a range to account for
"too" date.
Make params a date instead of a dateTime.
Update API and tests to account for param change.

stouffers committed:
add test.default.properties file

stouffers commented:
> Hey @stouffers, I can't find the `test.default.properties` file. Was it supposed to be committed or do I need to create it?

@KevinCaseiras just seeing this comment now. The file is supposed to be committed, but I neglected to do so earlier. Just pushed it up.

stouffers committed:
Change API key req header name

Use a more generic X-API-Key instead of X-BACHelp-API-Key

aa29cala reviewed (approved):
As per the dev meeting and phone calls this is good to go for an MVP
